[Ye Yint Paing Oo] iP

paing-oo opened the pull request:
# **DukePro**

> "Life is too complicated to not be orderly" - (Martha Stewart) [source](https://dansilvestre.com/productivity-quotes/)

- text-based
- easy to learn
- ~~FAST~~ _SUPER_ FAST to use

All you need to do is,

1. download it from [here](https://github.com/e0543517/ip/releases/tag/v0.1)
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you :100: 

And it is **FREE**!

Features:
- [x] Managing todo tasks
- [x] Managing deadlines
- [x] Managing events
- [ ] Reminders (coming soon)

If you are a Java Programmer, you can use it to practice java too. Here's the `main` method: 
```ruby
public static void main(String[] args) throws IOException {
        Duke duke = new Duke("data/duke.txt");
        duke.run();
    }
```

Commit:
Add Gradle support

paing-oo committed:
Implement an initial skeletal version of the Duke that simply greets the user and exits.

paing-oo committed:
Implement a skeletal version of Duke that starts by greeting the user, simply echos commands entered by the user, and exits when the user types bye.

paing-oo committed:
Added List class and Task class. Added the ability to store whatever text entered by the user and display them back to the user when requested.

paing-oo committed:
Add the ability to mark tasks as done and to change the status back to not done.

paing-oo committed:
Add support for tracking three types of tasks: ToDo Class, Event Class and Deadline Class.

paing-oo committed:
Automated Text UI Testing

paing-oo committed:
Added two exceptions handling class to deal with error such as incorrect inputs entered by user.

paing-oo committed:
Added support for deleting tasks from the list which also handles error when the delete user input is invalid.

paing-oo committed:
Added support for deleting tasks from the list which also handles an invalid delete user input.

paing-oo committed:
Added enums for Task types

paing-oo committed:
Add the feature to save the tasks in the hard disk and load the data from the hard disk when Duke starts up

paing-oo committed:
Store deadline date and time as a LocalDateTime object and print it in a certain format

paing-oo committed:
Merge branch 'branch-Level-7'

paing-oo committed:
Merge branch 'branch-Level-8'

# Conflicts:
#	src/main/java/Duke.java

paing-oo committed:
Make the code more OOP

Refractor the code to extract out closely related code as classes.

Extract the following classes:
* Ui : deals with interactions with user
* Storage: deals with loading tasks from the file and saving tasks in the file
* Parser: deals with making sense of the user command
* Command: deals with commands from user
* Create various child classes that inherits from Command Class

paing-oo committed:
Organise the classes into suitable packages

paing-oo committed:
Implement a way for users to find a task by searching for a keyword

paing-oo committed:
Add JavaDoc

paing-oo committed:
Implement appropriate Java Coding Standard

paing-oo committed:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/duke/Parser.java
#	src/main/java/duke/Storage.java
#	src/main/java/duke/command/DeleteCommand.java

paing-oo committed:
Change duke.txt

paing-oo committed:
Merge branch 'branch-A-JavaDoc'

# Conflicts:
#	src/main/java/duke/List.java
#	src/main/java/duke/Ui.java
#	src/main/java/duke/task/Deadline.java
#	src/main/java/duke/task/Task.java

talkintomato commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,84 @@
+     * @throws DukeException If user input is invalid.
+     */
+    public static Command parseUserInput(String userInput) throws DukeException {
+        checkIfUserInputValid(userInput);
```
Good abstraction of Invalid input

jltham replied:
May want to consider using switch-case here? Can help with specific error-checking as well!

talkintomato reviewed (commented):
Overall very clean code which abided by coding standards. Well written comments, method, and class names.

paing-oo committed:
Add JUnit Tests to test some methods in List class and Task class

paing-oo committed:
Create JAR file

paing-oo committed:
Merge branch 'branch-A-Gradle'

paing-oo committed:
implement checkstyle using checkstyle-IDEA plugin

paing-oo committed:
Implement checkstyle using checkstyle-IDEA plugin

paing-oo committed:
Merge branch 'branch-A-CheckStyle'

AdrianOngJJ commented on src/main/java/duke/command/ByeCommand.java:
```diff
@@ -4,11 +4,21 @@
+    /**
+     * Executes the ByeCommand and exits Duke.
+     *
+     * @param taskList TaskList.
+     * @param ui Ui.
+     * @param storage Storage.
```
Can provide a better description. If a description not needed, then we do not need to include it in the documentation

AdrianOngJJ reviewed (commented):
Nicely done!

paing-oo committed:
Add a GUI to the application

paing-oo committed:
Add a GUI to the application

paing-oo committed:
Merge branch 'branch-Level-10'

paing-oo committed:
Wrote some JavaDocs and fix stylecheck errors.

paing-oo committed:
Update appication main class name in build.gradle for GUI
Fix issue of list not showing fully in GUI

jltham commented on src/main/java/duke/List.java:
```diff
@@ -0,0 +1,143 @@
+/**
+ * Represents a list of the tasks added by the user.
+ */
+public class List {
```
May want to consider renaming List to TaskList? Especially since this class is specific to create a list of tasks.

jltham commented on src/main/java/duke/List.java:
```diff
@@ -0,0 +1,143 @@
+ */
+public class List {
+
+    private ArrayList<Task> arrayList;
```
Just nitpicking, but arrayList can be renamed as "tasksList" or "tasks" also? I think in general, more descriptive naming can help yourself and others read your code! 😃

jltham commented on src/main/java/duke/List.java:
```diff
@@ -0,0 +1,143 @@
+        switch (taskType) {
+        case TODO:
+            ToDo toDo = new ToDo(description);
+            arrayList.add(toDo);
```
I'm not sure if you have a reason behind assigning a new instance of ToDo to `toDo` then adding it to `arrayList`, but I thought this implementation can reduce your LoC!

```
arrayList.add(new ToDo(description));
```

May want to consider switching? Especially if `toDo` is not being using anywhere else!

jltham commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,114 @@
+            while (sc.hasNext()) {
+                line = sc.nextLine();
+                Character c = line.charAt(4);
+                if (c.equals('T')) {
```
May want to consider using switch-case here? If if-else chains help you organise your thoughts better then sure! I think switch-case can make the code look cleaner 😀

jltham commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,97 @@
+     * @param taskList Task List.
+     * @return Returns a String message.
+     */
+    public String showFindTask(List taskList) {
```
May want to consider shifting this method to be within List class? Having it within Ui class requires taskList to have a public method that exposes its private field of ArrayList, violating the abstraction barrier.

jltham commented on src/main/java/duke/command/AddCommand.java:
```diff
@@ -0,0 +1,80 @@
+     */
+    public AddCommand(Tasks taskType, String description, String time) {
+        super(false);
+        this.taskType = taskType;
```
May be simpler if you call 
```
this.AddCommand(taskType, description);
```
instead of typing out the whole of line 30 and 31?

jltham reviewed (commented):
LGTM! I really like how you organise your files and it's easy to follow 🥂 Just a few things I was nitpicking on, but no big deal! Good job 👍

paing-oo committed:
Add assertions

paing-oo committed:
Improve code quality

Some variable names were not following coding standards.

Change the variable names.

This improves readibility.

paing-oo committed:
Improve code quality

Method parseUserInput(String userInput) is too long.
Method load() is too long.

Apply SLAP for both methods and abstract out the codes into new methods.

This ensures that the code is more readable and follows good coding
standard.

paing-oo committed:
Merge pull request #2 from e0543517/branch-A-Assertions

Branch A assertions

paing-oo committed:
Merge branch 'master' of https://github.com/e0543517/ip

paing-oo committed:
Merge branch 'master' into branch-A-CodeQuality

# Conflicts:
#	src/main/java/duke/Parser.java

paing-oo committed:
Merge pull request #1 from e0543517/branch-A-CodeQuality

Branch A code quality

paing-oo committed:
Merge branch 'master' of https://github.com/e0543517/ip

paing-oo committed:
Implement C-Sort which sorts the tasks in the task list by their task
types and sorts deadline and event tasks by their local date time as
well

paing-oo committed:
Merge branch 'branch-BCD-Extension'

paing-oo committed:
Add JavaDocs

paing-oo committed:
Update README.md

paing-oo committed:
Update image size in dialog box

paing-oo committed:
Remove unused import

paing-oo committed:
Amend java doc

paing-oo committed:
Add files via upload

Add Ui.png

paing-oo committed:
Merge branch 'master' of https://github.com/e0543517/ip

paing-oo committed:
Update build.gradle dependencies to fix errors in jar file
